### ISSUE TITLE
MongoDB: Order of comparison in `where` clause affects the output

### ISSUE THREAD

**Issue by ludralph** [prisma/prisma#22263]
### Bug description

Changing the order of the filter in the Where clause affects the output for a simple `findMany()` call. This is based on this [Github Discussion](https://github.com/prisma/prisma/discussions/22253).

### How to reproduce

1. Given the schema below and using MongoDB connector.
2. insert records into the database and query for same. For example, see some records from the DB
```
[
  {
    data: null,
    objectId: '656f02b96d6a3b06cc1ff288',
    id: '1',
    createdAt: 2023-12-05T11:00:06.239Z,
    updatedAt: 2023-12-05T11:00:06.239Z,
    deletedAt: null
  },
  {
    data: null,
    objectId: '656f02ba6d6a3b06cc1ff289',
    id: '2',
    createdAt: 2023-12-05T11:00:10.321Z,
    updatedAt: 2023-12-05T11:00:10.321Z,
    deletedAt: null
  },
  {
    data: {
      createdAt: 2023-12-05T11:02:07.203Z,
      updatedAt: 2023-12-05T11:02:07.203Z,
      data1: 1
    },
    objectId: '656f0332f340065b906506fd',
    id: '3',
    createdAt: 2023-12-05T11:02:07.602Z,
    updatedAt: 2023-12-05T11:02:07.602Z,
    deletedAt: null
  }
]
```
3. When i query like this
```js
 const res = await prisma.account.findMany({
    where: {
      deletedAt: null,
      data:{
        isNot: null
      },
    },
  });
```
It returns the result set below:
```
[
  {
    data: {
      createdAt: 2023-12-05T11:02:07.203Z,
      updatedAt: 2023-12-05T11:02:07.203Z,
      data1: 1
    },
    objectId: '656f0332f340065b906506fd',
    id: '3',
    createdAt: 2023-12-05T11:02:07.602Z,
    updatedAt: 2023-12-05T11:02:07.602Z,
    deletedAt: null
  }
]
```
4.  if I change the order of the parameters like shown:
```js
const res = await prisma.account.findMany({
    where: {
      data:{
        isNot: null
      },
      deletedAt: null,

    },
  });
```
Notice that `deletedAt` is now after `data` property. It returns an empty resultset [ ]. 
-->

### Expected behavior

I would expect the result of both queries to return the same output.

### Prisma information

```prisma
type Data {
  createdAt DateTime
  updatedAt DateTime
  data1     Int
}

model Account {
  objectId String @id @default(auto()) @map("_id") @db.ObjectId
  id       String @unique

  data Data?

  createdAt DateTime? @default(now())
  updatedAt DateTime? @updatedAt
  deletedAt DateTime?

  @@map("accounts")
}
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS:  macOS
- Database:  MongoDB
- Node.js version: v20.2.0

### Prisma Version

```
prisma                  : 5.6.0
@prisma/client          : 5.6.0
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v20.2.0
Query Engine (Node-API) : libquery-engine e95e739751f42d8ca026f6b910f5a2dc5adeaeee (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli e95e739751f42d8ca026f6b910f5a2dc5adeaeee (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.6.0-32.e95e739751f42d8ca026f6b910f5a2dc5adeaeee
Default Engines Hash    : e95e739751f42d8ca026f6b910f5a2dc5adeaeee
Studio                  : 0.495.0
Preview Features        : fullTextSearch
```

**Comment by nurul3101** [comment]
A sample reference for the generated underlying MongoDB queries

```ts
await prisma.mainmodel.count({where: {AND: [{roommate: {isNot: null}}, {booking: {is: {name: "TEST"}}}]}})
prisma:query prisma.mainmodel.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $and: [ { $not: { $eq: [ "$roommate", null, ], }, }, { $ne: [ "$roommate", "$$REMOVE", ], }, ], }, ], }, { $and: [ { $ne: [ "$booking.name", { $literal: "TEST", }, ], }, { $ne: [ "$booking.name", "$$REMOVE", ], }, ], }, ], }, }, }, { $group: { _id: null, count_all: { $sum: 1, }, }, }, ])


await prisma.mainmodel.count({where: {AND: [{booking: {is: {name: "TEST"}}}, {roommate: {isNot: null}}]}})
prisma:query prisma.mainmodel.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $eq: [ "$booking.name", { $literal: "TEST", }, ], }, { $ne: [ "$booking.name", "$$REMOVE", ], }, ], }, { $and: [ { $and: [ { $not: { $eq: [ "$roommate", null, ], }, }, { $ne: [ "$roommate", "$$REMOVE", ], }, ], }, ], }, ], }, }, }, { $group: { _id: null, count_all: { $sum: 1, }, }, }, ])
```

**Comment by janpio** [comment]
Related issue: https://github.com/prisma/prisma/issues/22970

**Comment by SevInf** [comment]
I can confirm the issue on 5.6.1 and i also can confirm that it was fixed in Prisma 5.11 (https://github.com/prisma/prisma-engines/pull/4754).

I am going to close this issue now, but feel free to comment on if it still happens to you on Prisma 5.11 or later.
